### DISCUSSION TITLE
`set()` with function does not throw type errors when return data is incorrect

### DISCUSSION THREAD

**Discussion by mikaelkarlsson-se** [pmndrs/zustand#2879] [Bug report] [answered]
### Bug Description

Hello!
I've discovered that when using a function in `set()` there will be no TypeScript errors thrown if the new data has an incorrect structure. Since the type is `set: (partial: ExampleStore | Partial<ExampleStore> | ((state: ExampleStore) => ExampleStore | Partial<ExampleStore>), replace?: false) => void` it should throw an error as far as I understand as it expects a certain return type. However if I set the function return type explicitly it will throw an error.

A small example is this (available at StackBlitz, see reproduction link):
```
import { create } from 'zustand';

type State = {
  firstProperty: string;
  secondProperty: number;
  someThingNested: {
    thirdProperty: boolean;
  };
};

type ExampleStore = {
  exampleState: State;
  updateState: (newState: State) => void;
};

export const useExampleStore = create<ExampleStore>((set) => ({
  exampleState: {
    firstProperty: 'I am just an example',
    secondProperty: 0,
    someThingNested: {
      thirdProperty: true,
    },
  },
  updateState: (newState) =>
    // This does not throw a type error
    set((prevState) => ({
      // This throw a type error
      // set((prevState): Partial<ExampleStore> => ({
      exampleState: {
        ...prevState,
        ...newState,
        somePropertyThatDoesNotExistInTheType: 'I should throw a type error',
        someThingNested: {
          ...prevState,
          ...newState.someThingNested,
        },
      },
    })),
}));
```

And since I'm writing anyway, just want to grasp the opportunity and say thanks for this nice library! 🤝 

### Reproduction Link

https://stackblitz.com/edit/vitejs-vite-igq5xy?file=src%2FexampleStore.ts

**Comment by dai-shi** [comment]
It doesn't show the error in TS playground.
https://tsplay.dev/WoZE9m



> And since I'm writing anyway, just want to grasp the opportunity and say thanks for this nice library! 🤝

Thanks!!

**Reply by mikaelkarlsson-se** [reply]
Not sure if you mean that it seems to work as intended or not, or just a reproduction in another environment? 😄 
The playground code has the same flaw as my Stackblitz link. `somePropertyThatDoesNotExistInTheType` should always throw a type error since there's an expected return type where it is not included, but it only does throw an error when adding `(prevState): ExampleStore =>`.

**Reply by dai-shi** [reply]
Oh, I think I misunderstood the issue description completely the opposite. You mean it should cause TS error.
To my knowledge, this is impossible with the current TypeScript. Or, I'm not a TS wizard yet.

**Reply by mikaelkarlsson-se** [reply]
I'm not a TS wizard either so I might be misunderstanding the types or it's perhaps some kind of TS limitation that I don't know about. In my mind, by reading the type for `set()` which is a union containing `(state: ExampleStore) => ExampleStore | Partial<ExampleStore>` as one of the possible types, I expect TS to make sure the return from `(prevState) => {...} ` is `ExampleStore | Partial<ExampleStore>` but it doesn't. However it does if I add the expected return type `(prevState): ExampleStore | Partial<ExampleStore> => {...}` which I would say is the same type definition, but perhaps I'm not correct about this? 🤔

**Reply by dai-shi** [reply]
Here's an example without zustand: https://tsplay.dev/m3DeAW

```ts
type State = {
  foo: string;
};

type Fn = () => State;

declare function f(fn: Fn): void;

f(() => ({ foo: "aaa" }));
f(() => ({ foo: "aaa", bar: "bbb" }));
f(() => ({ foo: 111 }));
```

**Reply by mikaelkarlsson-se** [reply, marked as answer]
Hmm, seems to be a limitation or something I haven't understood yet in the world of TS then. Closing this since it's obviously not a bug in zustand. Thanks for the help @dai-shi 🤝
